dbms/src/DataTypes/DataTypeString.cpp

```diff
@@ -310,4 +317,203 @@ void registerDataTypeString(DataTypeFactory & factory)
     factory.registerSimpleDataType("LONGBLOB", creator, DataTypeFactory::CaseInsensitive);
 }
 
+namespace
+{
+
+using Offset = ColumnString::Offset;
+
+// Returns <offsets_stream, chars_stream>.
+template <typename B, typename G>
+std::pair<B *, B *> getStream(const G & getter, IDataType::SubstreamPath & path)
+{
+    auto * chars_stream = getter(path);
+    path.emplace_back(IDataType::Substream::StringSizes);
+    auto * offsets_stream = getter(path);
+    return {offsets_stream, chars_stream};
+}
+
+PaddedPODArray<Offset> offsetToStrSize(
+    const ColumnString::Offsets & chars_offsets,
+    const size_t begin,
+    const size_t end)
+{
+    assert(!chars_offsets.empty());
+    // The class PODArrayBase ensure chars_offsets[-1] is well defined as 0.
+    // For details, check the `pad_left` argument in PODArrayBase.
+    // In the for loop code below, when `begin` and `i` are 0:
+    // str_sizes[0] = chars_offsets[0] - chars_offsets[-1];
+    assert(chars_offsets[-1] == 0);
+
+    PaddedPODArray<Offset> str_sizes(end - begin);
+    auto chars_offsets_pos = chars_offsets.begin() + begin;
+
+    // clang-format off
+    #pragma clang loop vectorize(enable)
+    // clang-format on
+    for (ssize_t i = 0; i < static_cast<ssize_t>(str_sizes.size()); ++i)
+    {
+        str_sizes[i] = chars_offsets_pos[i] - chars_offsets_pos[i - 1];
+    }
```

Comment:
How about using a prefix sum like the original behavior? If so, the `strSizeToOffset` can just use a simple memcpy.

Reply:
What you describe is like the `position_independent_encoding == false` in `ColumnArray`. Such behavior will bring problem when someone want to deserialize then append the data into a column with existing data.
`CHBlockChunkDecodeAndSquash` now will try to merge the deserialized data into a larger block, so it will cause issue like https://github.com/pingcap/tiflash/pull/9477.

In order to be safe, we always use `position_independent_encoding == true` in TiFlash now.

Reply:
Got it.